### DISCUSSION TITLE
Setting RTC from ntp

### DISCUSSION THREAD

**Discussion by ve3sjk** [Makuna/Rtc#162] [General]
I have esp32 ntp setup and would like to set the ds3231 time using the time from ntp but I am drawing a blank on what format i have to convert the ntp struc to before passing it to Rtc.SetDateTime()

Ntp is setup and working using an example that uses this to print the output.

`void printLocalTime(){
  struct tm timeinfo;
  if(!getLocalTime(&timeinfo)){
    Serial.println("Failed to obtain time");
    return;
  }
  Serial.println(&timeinfo, "%A, %B %d %Y %H:%M:%S");
  Serial.print("Day of week: ");
  Serial.println(&timeinfo, "%A");
  Serial.print("Month: ");
  Serial.println(&timeinfo, "%B");
  Serial.print("Day of Month: ");
  Serial.println(&timeinfo, "%d");
  Serial.print("Year: ");
  Serial.println(&timeinfo, "%Y");
  Serial.print("Hour: ");
  Serial.println(&timeinfo, "%H");
  Serial.print("Hour (12 hour format): ");
  Serial.println(&timeinfo, "%I");
  Serial.print("Minute: ");
  Serial.println(&timeinfo, "%M");
  Serial.print("Second: ");
  Serial.println(&timeinfo, "%S");

  Serial.println("Time variables");
  char timeHour[3];
  strftime(timeHour,3, "%H", &timeinfo);
  Serial.println(timeHour);
  char timeWeekDay[10];
  strftime(timeWeekDay,10, "%A", &timeinfo);
  Serial.println(timeWeekDay);
  Serial.println();
} `

**Comment by Makuna** [comment]
[FAQ #3- How do I set the time from NTP Server or with a Epoch time?](https://github.com/Makuna/Rtc/wiki/FAQ-%233)

**Reply by Makuna** [reply]
NOTE:  I created an issue to specifically to add a method to InitWithNtp(uint64_t secondsFrom1900);
So at some point it will be easier.

**Reply by ve3sjk** [reply]
I have it almost working based on your first reply to my question I am not sure i am using the right symbols to get the code to look proper in the discussion area please correct me so i can format it right if it isn't

I am using this code to get the current epoch time from NTP using the builtin NTP functions on the esp32
```

// Function that gets current epoch time
unsigned long getTime() {
  time_t now;
  struct tm timeinfo;
  if (!getLocalTime(&timeinfo)) {
    //Serial.println("Failed to obtain time");
    return(0);
  }
  time(&now);
  return now;
}
```

and using this function to update the ds3231 the print statements are for debugging
```

void RTC_Update(){
  RtcDateTime now = Rtc.GetDateTime();//rtc
  Serial.println("************************************");
  Serial.println("RTC time before update in RTC_Update");
  printDateTime(now);
  Serial.println("NTP time before update in RTC_Update");  
  printLocalTime();
  Serial.println();      
  RtcDateTime timeToSet;
  timeToSet.InitWithEpoch32Time(getTime());
  Serial.println("timeToSet value before update in RTC_Update");
  Rtc.SetDateTime(timeToSet);
  printDateTime(timeToSet);
  Serial.println();
  Serial.println("RTC time after update in RTC_Update");
  now = Rtc.GetDateTime();
  printDateTime(now); 
  Serial.println("Set RTC to NTP complete in RTC_Update");
  Serial.println("*************************************");   
  }
```


Based on this output on the serial monitor it is setting the RTC now but it appears to be setting it to UTC time and ignoring the timezone despite the fact that this code from NTP does print the right time

```
void printLocalTime(){
  struct tm timeinfo;
  if(!getLocalTime(&timeinfo)){
    Serial.println("Failed to obtain time 1");
    return;
  }
  Serial.println(&timeinfo, "%A, %B %d %Y %H:%M:%S zone %Z %z ");
}
```

I would have thought the correct time with timezone adjustment would have been passed by the getime() function since its near identical to the code in printlocal time.

Here is the output on the serial monitor of what it is doing.

```
19:32:39.106 -> Booting
19:32:39.204 -> [WiFi-event] event: 0
19:32:39.204 -> [WiFi-event] event: 2
19:32:39.204 -> [WiFi-event] event: 4
19:32:41.282 -> [WiFi-event] event: 7
19:32:41.315 -> WiFi connected
19:32:41.315 -> IP address: 
19:32:41.315 -> 192.168.1.8
19:32:41.315 -> Connecting to MQTT...
19:32:41.315 -> Connected to MQTT.
19:32:41.315 -> Session present: 0
19:32:41.315 -> topic home/nb/node04/dev99 value:NODE 4 WAKEUP: ESP32_04
19:32:41.412 -> Setting up time
19:32:41.477 ->   Got the time from NTP
19:32:41.477 ->   Setting Timezone to EST5EDT,M3.2.0,M11.1.0
19:32:41.477 -> Thursday, March 16 2023 19:32:41 zone EDT -0400 
19:32:41.477 -> 
19:32:41.477 -> 
19:32:41.477 -> RTC time before update in setup
19:32:41.477 -> 03/16/2023 23:32:40
19:32:41.477 -> ************************************
19:32:41.477 -> RTC time before update in RTC_Update
19:32:41.477 -> 03/16/2023 23:32:40
19:32:41.477 -> NTP time before update in RTC_Update
19:32:41.477 -> Thursday, March 16 2023 19:32:41 zone EDT -0400 
19:32:41.477 -> 
19:32:41.477 -> timeToSet value before update in RTC_Update
19:32:41.510 -> 03/16/2023 23:32:41
19:32:41.510 -> 
19:32:41.510 -> RTC time after update in RTC_Update
19:32:41.510 -> 03/16/2023 23:32:41
19:32:41.510 -> Set RTC to NTP complete in RTC_Update
19:32:41.510 -> *************************************
19:32:43.489 -> RTC time after update in setup
19:32:43.489 -> 03/16/2023 23:32:43
19:32:43.489 -> 
19:32:43.489 -> Ready
19:32:43.489 -> IP address: 192.168.1.8
19:32:43.522 -> I2C device found at address 0x21
19:32:45.535 -> I2C device found at address 0x27
19:32:47.575 -> I2C device found at address 0x3C
19:32:49.619 -> I2C device found at address 0x57
19:32:51.621 -> I2C device found at address 0x68
19:32:53.657 -> I2C device found at address 0x77
19:32:55.700 -> done
19:32:55.700 -> 
19:32:58.521 -> 
19:32:58.521 -> BSEC library version 1.4.8.0
```

**Comment by Makuna** [comment]
https://github.com/Makuna/Rtc/pull/164

Note that RtcDateTime doesn't even consider/track timezone as the RTC chips don't either.  
Most commercial code for time will leave time in UTC and only convert to local when it gets displayed or convert from it directly from user input; but they tend to keep official time in UTC.  But this is Arduino, so it's left up to the user to standardize in their sketch.  But don't assume a printLocal doesn't adjust for timezone for you; otherwise why name it with "local" versus just printTime.

**Reply by Makuna** [reply]
Note that the new method names should make it more clear what to call.
```
    uint32_t Ntp32Time() const
    void InitWithNtp32Time(uint32_t secondsSince1900)

    uint32_t Unix32Time() const
    void InitWithUnix32Time(uint32_t secondsSince1970)
```

**Reply by ve3sjk** [reply]
Thanks again for getting me on the right track on this section of code, i agree I am going to leave and store to mysql the timestamp in UTC and do the conversions externally, if i decided to do it in my local time i suppose it would be a matter of getting the %z variable from the NTP struct and converting it to seconds, in my case -4 hours would subtract 14400 seconds from the epoch time in the gettime() function before returning the value to the call. Now I need to redo all my alarm routines as they were written with a different ds3232 library before i switched to yours, from the examples I don't think is going to be very difficult.
